File: sample-servers/authorization-server/src/main/java/com/dm/server/authorization/converter/DmAccessTokenConverter.java
package com.dm.server.authorization.converter;

import com.dm.security.core.userdetails.UserDetailsDto;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;
import org.xyyh.authorization.core.OAuth2Authentication;
import org.xyyh.authorization.core.OAuth2ServerAccessToken;
import org.xyyh.authorization.endpoint.converter.DefaultAccessTokenConverter;

import java.util.Map;
import java.util.Objects;

import static org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionClaimNames.SUBJECT;

@Component
public class DmAccessTokenConverter extends DefaultAccessTokenConverter {
    @Override
    public Map<String, Object> toAccessTokenIntrospectionResponse(OAuth2ServerAccessToken token, OAuth2Authentication authentication) {
        Object principal = authentication.getPrincipal();
        Map<String, Object> response = super.toAccessTokenIntrospectionResponse(token, authentication);
        if (principal instanceof String) {
            response.put(SUBJECT, principal);
        } else if (principal instanceof UserDetailsDto) {
            Long userId = ((UserDetailsDto) principal).getId();
            if (Objects.nonNull(userId)) {
                response.put(SUBJECT, String.valueOf(userId));
            }
        } else if (principal instanceof UserDetails) {
            response.put(SUBJECT, ((UserDetails) principal).getUsername());
        }
        return response;
    }
}

File: sample-servers/authorization-server/src/main/kotlin/com/dm/server/authorization/converter/DmAccessTokenConverter.kt
package com.dm.server.authorization.converter

import com.dm.security.core.userdetails.UserDetailsDto
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionClaimNames.SUBJECT
import org.springframework.stereotype.Component
import org.xyyh.oidc.core.OidcAuthentication
import org.xyyh.oidc.core.OAuth2ServerAccessToken
import org.xyyh.oidc.endpoint.converter.DefaultAccessTokenConverter


@Component
class DmAccessTokenConverter : DefaultAccessTokenConverter() {
    override fun toAccessTokenIntrospectionResponse(
        token: OAuth2ServerAccessToken?,
        authentication: OidcAuthentication
    ): Map<String, Any> {
        val details = authentication.principal
        val response = super.toAccessTokenIntrospectionResponse(token, authentication)
        response["authorities"] = authentication.authorities
        when (details) {
            is String -> {
                response[SUBJECT] = details
            }
            is UserDetailsDto -> {
                details.id?.let {
                    response[SUBJECT] = it.toString()
                }
                // 添加用户名
                response["fullname"] = details.fullname
            }
            is UserDetails -> {
                response[SUBJECT] = details.username
            }
        }
        return response
    }
}
